File: internal/boundary/options.go
package boundary

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"reflect"
	"strings"

	markdownviewer "github.com/sriannamalai/markdownviewer"
	"github.com/sriannamalai/markdownviewer/parser"
)

// options is the version-1 options JSON accepted by every exported
// function. Fields irrelevant to an operation (e.g. theme for mdv_parse,
// or parser for mdv_render_doc, which consumes an already-parsed
// document) are decoded and ignored by that operation.
type options struct {
	Version        *int              `json:"version"`
	Theme          string            `json:"theme"`
	Fragment       bool              `json:"fragment"`
	AllowRawHTML   bool              `json:"allowRawHTML"`
	Mermaid        bool              `json:"mermaid"`
	Math           bool              `json:"math"`
	Highlighting   bool              `json:"highlighting"`
	MaxWidth       string            `json:"maxWidth"`
	SourceMap      bool              `json:"sourceMap"`
	ThemeOverrides map[string]string `json:"themeOverrides"`
	Stylesheet     string            `json:"stylesheet"`
	ExtraCSS       string            `json:"extraCss"`
	CodeHeader     bool              `json:"codeHeader"`
	HeadingAnchors bool              `json:"headingAnchors"`
	Parser         *parserOptions    `json:"parser"`
}

// parserOptions is the nested "parser" object: a 1:1 JSON mirror of
// parser.Config's syntax-extension toggles. Absent (or JSON null) means
// "library default" — parser.Default(), every extension on — with
// byte-identical output to older payloads that had no such field.
//
// commonmarkOnly selects parser.CommonMarkOnly() (the zero Config, no
// extensions) as the base instead of parser.Default(). The per-extension
// fields are tristate overrides applied on top of that base: absent/null
// keeps the base's setting, true enables, false disables — so
// {"commonmarkOnly": true, "tables": true} is pure CommonMark plus GFM
// tables, and {"wikiLinks": false} is the full default set minus
// wiki-links. Note the nested "math" toggle is parse-side ([$x$] syntax
// recognition); the top-level "math" option is render-side (KaTeX).
type parserOptions struct {
	CommonmarkOnly  bool  `json:"commonmarkOnly"`
	Tables          *bool `json:"tables"`
	Strikethrough   *bool `json:"strikethrough"`
	TaskLists       *bool `json:"taskLists"`
	Linkify         *bool `json:"linkify"`
	Footnotes       *bool `json:"footnotes"`
	DefinitionLists *bool `json:"definitionLists"`
	FrontMatter     *bool `json:"frontMatter"`
	Emoji           *bool `json:"emoji"`
	WikiLinks       *bool `json:"wikiLinks"`
	Math            *bool `json:"math"`
	Admonitions     *bool `json:"admonitions"`
}

// toConfig folds the nested parser object into a parser.Config.
func (p *parserOptions) toConfig() parser.Config {
	cfg := parser.Default()
	if p.CommonmarkOnly {
		cfg = parser.CommonMarkOnly()
	}
	for dst, v := range map[*bool]*bool{
		&cfg.Tables:          p.Tables,
		&cfg.Strikethrough:   p.Strikethrough,
		&cfg.TaskLists:       p.TaskLists,
		&cfg.Linkify:         p.Linkify,
		&cfg.Footnotes:       p.Footnotes,
		&cfg.DefinitionLists: p.DefinitionLists,
		&cfg.FrontMatter:     p.FrontMatter,
		&cfg.Emoji:           p.Emoji,
		&cfg.WikiLinks:       p.WikiLinks,
		&cfg.Math:            p.Math,
		&cfg.Admonitions:     p.Admonitions,
	} {
		if v != nil {
			*dst = *v
		}
	}
	return cfg
}

// jsonKeys returns the exact JSON key of every field of the struct type
// v, derived from the struct tags so it cannot drift from the decoder.
func jsonKeys(v any) map[string]bool {
	keys := map[string]bool{}
	t := reflect.TypeOf(v)
	for i := 0; i < t.NumField(); i++ {
		name, _, _ := strings.Cut(t.Field(i).Tag.Get("json"), ",")
		keys[name] = true
	}
	return keys
}

// knownOptionKeys / knownParserOptionKeys hold the exact JSON keys of
// options and its nested parser object. encoding/json matches field names
// case-insensitively even with DisallowUnknownFields, so strict decoding
// additionally requires an exact-case key match against these sets (e.g.
// "extraCSS" must be rejected, not silently folded into "extraCss" —
// likewise "parser.wikilinks" vs "parser.wikiLinks").
var (
	knownOptionKeys       = jsonKeys(options{})
	knownParserOptionKeys = jsonKeys(parserOptions{})
)

func defaultOptions() options {
	return options{Theme: "auto", Mermaid: true, Math: true, Highlighting: true, HeadingAnchors: true}
}

// decodeOptions strictly decodes the options JSON over the defaults.
// nil, empty, or all-whitespace input yields the defaults.
func decodeOptions(data []byte) (options, error) {
	o := defaultOptions()
	if len(bytes.TrimSpace(data)) == 0 {
		return o, nil
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err == nil {
		// Malformed/trailing input is left for the strict decode below to
		// report; this pass only enforces exact-case key names.
		for k := range raw {
			if !knownOptionKeys[k] {
				return options{}, fmt.Errorf("options: unknown field %q", k)
			}
		}
		// The nested "parser" object gets the same exact-case treatment.
		// Its value arrives as a json.RawMessage here; re-unmarshalling it
		// as a key map succeeds for an object (and for JSON null, which
		// yields a nil map — same as an absent field). Any other value
		// kind is rejected up front with a clearer message than the strict
		// decoder's type error.
		if p, ok := raw["parser"]; ok {
			var nested map[string]json.RawMessage
			if err := json.Unmarshal(p, &nested); err != nil {
				return options{}, fmt.Errorf("options: \"parser\" must be a JSON object")
			}
			for k := range nested {
				if !knownParserOptionKeys[k] {
					return options{}, fmt.Errorf("options: unknown field %q", "parser."+k)
				}
			}
		}
	}
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&o); err != nil {
		return options{}, fmt.Errorf("options: %w", err)
	}
	if _, err := dec.Token(); err != io.EOF {
		return options{}, fmt.Errorf("options: trailing data after the options object")
	}
	if o.Version != nil && *o.Version != 1 {
		return options{}, fmt.Errorf("options: unsupported version %d (want 1)", *o.Version)
	}
	return o, nil
}

func (o options) toFacadeOptions(resolver markdownviewer.Resolver) []markdownviewer.Option {
	var opts []markdownviewer.Option
	if resolver != nil {
		opts = append(opts, markdownviewer.WithResolver(resolver))
	}
	if o.Theme != "" {
		opts = append(opts, markdownviewer.WithTheme(o.Theme))
	}
	if o.Fragment {
		opts = append(opts, markdownviewer.Fragment())
	}
	if o.AllowRawHTML {
		opts = append(opts, markdownviewer.AllowRawHTML())
	}
	if !o.Mermaid {
		opts = append(opts, markdownviewer.DisableMermaid())
	}
	if !o.Math {
		opts = append(opts, markdownviewer.DisableMath())
	}
	if !o.Highlighting {
		opts = append(opts, markdownviewer.DisableHighlighting())
	}
	if !o.HeadingAnchors {
		opts = append(opts, markdownviewer.DisableHeadingAnchors())
	}
	// Parse-time only: Render/RenderTo parse with this config; RenderDoc
	// receives an already-parsed tree, so WithParserConfig is a no-op
	// there (the field is decoded and ignored, like theme for mdv_parse).
	if o.Parser != nil {
		opts = append(opts, markdownviewer.WithParserConfig(o.Parser.toConfig()))
	}
	if o.SourceMap {
		opts = append(opts, markdownviewer.WithSourceMap())
	}
	if len(o.ThemeOverrides) > 0 {
		opts = append(opts, markdownviewer.WithThemeOverrides(o.ThemeOverrides))
	}
	// After WithThemeOverrides: it replaces the whole override map, and
	// WithMaxWidth sets a key inside it (see the facade godoc).
	if o.MaxWidth != "" {
		opts = append(opts, markdownviewer.WithMaxWidth(o.MaxWidth))
	}
	if o.Stylesheet != "" {
		opts = append(opts, markdownviewer.WithStylesheet(o.Stylesheet))
	}
	if o.ExtraCSS != "" {
		opts = append(opts, markdownviewer.WithExtraCSS(o.ExtraCSS))
	}
	if o.CodeHeader {
		opts = append(opts, markdownviewer.WithCodeHeader())
	}
	return opts
}

File: flutter/mdviewer/lib/src/options.dart
import 'dart:convert';

/// Resolver callback kind. The index IS the ABI int used by
/// `mdv_render_r` / `mdv_render_doc_r` (`0` link, `1` image, `2`
/// wiki-link) — see `ffi/README.md`'s "Resolver callback" table.
enum MdvResolveKind { link, image, wikiLink }

/// Host resolver callback for [MdvOptions.resolver]. Return a resolved
/// URL, or `null` to decline (default resolution applies). Called
/// synchronously on the calling thread during a render; a thrown error
/// is recorded, declines the remainder of the render, and is rethrown as
/// a boundary exception after the native call returns (see
/// `_callWithResolver` in `mdviewer_base.dart`).
typedef MdvResolver = String? Function(MdvResolveKind kind, String target);

/// Typed, strict-by-construction mirror of the boundary's nested
/// `"parser"` options object: which Markdown syntax extensions the parse
/// enables (a 1:1 mirror of the Go `parser.Config`).
///
/// Omitting [MdvOptions.parser] entirely means "library default" — every
/// extension on. [commonmarkOnly] starts from pure CommonMark (no
/// extensions) instead; the per-extension fields are tristate overrides
/// on top of that base: `null` (unset) keeps the base's setting, `true`
/// enables, `false` disables. So `MdvParserOptions(wikiLinks: false)`
/// renders `[[x]]` as literal text, and
/// `MdvParserOptions(commonmarkOnly: true, tables: true)` is CommonMark
/// plus GFM tables.
///
/// Parse-time only: affects `render` and `parse`; `renderDoc` receives
/// an already-parsed document, so the boundary decodes and ignores it
/// there. Note [math] gates `$x$` syntax recognition at parse time,
/// while [MdvOptions.math] gates KaTeX rendering.
class MdvParserOptions {
  const MdvParserOptions({
    this.commonmarkOnly,
    this.tables,
    this.strikethrough,
    this.taskLists,
    this.linkify,
    this.footnotes,
    this.definitionLists,
    this.frontMatter,
    this.emoji,
    this.wikiLinks,
    this.math,
    this.admonitions,
  });

  final bool? commonmarkOnly;
  final bool? tables;
  final bool? strikethrough;
  final bool? taskLists;
  final bool? linkify;
  final bool? footnotes;
  final bool? definitionLists;
  final bool? frontMatter;
  final bool? emoji;
  final bool? wikiLinks;
  final bool? math;
  final bool? admonitions;

  /// The nested `"parser"` JSON object: only explicitly-set (non-null)
  /// fields serialize, under their exact boundary key names. An empty
  /// map is still meaningful ("parser present, all defaults") and
  /// behaves identically to omitting [MdvOptions.parser].
  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{};
    if (commonmarkOnly != null) map['commonmarkOnly'] = commonmarkOnly;
    if (tables != null) map['tables'] = tables;
    if (strikethrough != null) map['strikethrough'] = strikethrough;
    if (taskLists != null) map['taskLists'] = taskLists;
    if (linkify != null) map['linkify'] = linkify;
    if (footnotes != null) map['footnotes'] = footnotes;
    if (definitionLists != null) map['definitionLists'] = definitionLists;
    if (frontMatter != null) map['frontMatter'] = frontMatter;
    if (emoji != null) map['emoji'] = emoji;
    if (wikiLinks != null) map['wikiLinks'] = wikiLinks;
    if (math != null) map['math'] = math;
    if (admonitions != null) map['admonitions'] = admonitions;
    return map;
  }
}

/// Typed, strict-by-construction options for [Mdviewer.render],
/// [Mdviewer.parse], and [Mdviewer.renderDoc].
///
/// Every field is typed and named; there is no free-form map, so no
/// unknown key can ever reach the boundary. [toJson] serializes only the
/// fields that were explicitly set (non-null) — an omitted field means
/// "library default", matching the boundary's own NULL/omitted-field
/// semantics (see `ffi/README.md`'s Options JSON table). [resolver] is
/// consumed entirely on the Dart side of the call and never serializes.
class MdvOptions {
  const MdvOptions({
    this.theme,
    this.fragment,
    this.allowRawHTML,
    this.mermaid,
    this.math,
    this.highlighting,
    this.maxWidth,
    this.sourceMap,
    this.themeOverrides,
    this.stylesheet,
    this.extraCss,
    this.codeHeader,
    this.headingAnchors,
    this.parser,
    this.resolver,
  });

  final String? theme;
  final bool? fragment;
  final bool? allowRawHTML;
  final bool? mermaid;
  final bool? math;
  final bool? highlighting;
  final String? maxWidth;
  final bool? sourceMap;
  final Map<String, String>? themeOverrides;
  final String? stylesheet;

  /// Extra CSS appended AFTER whatever base styling applied (base+theme,
  /// or [stylesheet] when set — whose replace semantics are unchanged).
  /// Full-page assembly only; no effect in fragment mode.
  final String? extraCss;

  /// When `true`, code blocks are wrapped in an `md-code` container with a
  /// language-label + Copy-button header (full pages also get the inline
  /// clipboard JS). Default `false`: output unchanged.
  final bool? codeHeader;

  /// Slug `id` attributes on headings (`<h1 id="...">`). Library default
  /// `true`; set `false` to omit them (intra-page `#fragment` links to
  /// headings stop resolving). Render-time: applies to `render` and
  /// `renderDoc`.
  final bool? headingAnchors;

  /// Nested parser configuration (which syntax extensions the parse
  /// enables); see [MdvParserOptions]. `null` means library default —
  /// every extension on.
  final MdvParserOptions? parser;

  final MdvResolver? resolver;

  /// The `opts_json` payload for this options set, or `null` when every
  /// non-[resolver] field is null — the NULL-options path, which asks
  /// the boundary for its own defaults rather than sending `{}`.
  String? toJson() {
    final map = <String, dynamic>{};
    if (theme != null) map['theme'] = theme;
    if (fragment != null) map['fragment'] = fragment;
    if (allowRawHTML != null) map['allowRawHTML'] = allowRawHTML;
    if (mermaid != null) map['mermaid'] = mermaid;
    if (math != null) map['math'] = math;
    if (highlighting != null) map['highlighting'] = highlighting;
    if (maxWidth != null) map['maxWidth'] = maxWidth;
    if (sourceMap != null) map['sourceMap'] = sourceMap;
    if (themeOverrides != null) map['themeOverrides'] = themeOverrides;
    if (stylesheet != null) map['stylesheet'] = stylesheet;
    if (extraCss != null) map['extraCss'] = extraCss;
    if (codeHeader != null) map['codeHeader'] = codeHeader;
    if (headingAnchors != null) map['headingAnchors'] = headingAnchors;
    if (parser != null) map['parser'] = parser!.toJson();
    if (map.isEmpty) return null;
    return jsonEncode(map);
  }
}
